processors/ratelimiter/factory.go

```diff
@@ -0,0 +1,87 @@
+package ratelimiter
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"time"
+
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/consumer"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const (
+	typeStr                         = "hypertrace_ratelimiter"
+	defaultServiceHost              = "127.0.0.1"
+	defaultServicePort              = uint16(8081)
+	defaultDomain                   = "collector"
+	defaultDomainSoftLimitThreshold = uint32(100000) // Soft limit kicks in when limit remaining under 100k
+	defaultHeaderName               = "x-tenant-id"
+	defaultTimeoutMillis            = uint32(1000) // 1 second
+)
+
+// NewFactory creates a factory for the ratelimit processor.
+func NewFactory() component.ProcessorFactory {
+	return component.NewProcessorFactory(
+		typeStr,
+		createDefaultConfig,
+		component.WithTracesProcessor(createTraceProcessor, component.StabilityLevelStable),
+	)
+}
+
+func createDefaultConfig() config.Processor {
+	return &Config{
+		ProcessorSettings: config.NewProcessorSettings(
+			config.NewComponentID(typeStr),
+		),
+		RateLimitServiceHost:         defaultServiceHost,
+		RateLimitServicePort:         defaultServicePort,
+		Domain:                       defaultDomain,
+		DomainSoftRateLimitThreshold: defaultDomainSoftLimitThreshold,
+		TenantIDHeaderName:           defaultHeaderName,
+	}
+}
+
+func createTraceProcessor(
+	ctx context.Context,
+	params component.ProcessorCreateSettings,
+	cfg config.Processor,
+	nextConsumer consumer.Traces,
+) (component.TracesProcessor, error) {
+	pCfg := cfg.(*Config)
+	rateLimitServiceClient, err := getRateLimitServiceClient(ctx, pCfg.RateLimitServiceHost, pCfg.RateLimitServicePort, pCfg.RateLimitServiceTimeoutMillis, params)
+	if err != nil {
+		params.Logger.Error("failed to connect to rate limit service ", zap.Error(err))
+		return nil, err
+	}
+	processor := &processor{
+		rateLimitServiceClient:   rateLimitServiceClient,
+		domain:                   pCfg.Domain,
+		domainSoftLimitThreshold: pCfg.DomainSoftRateLimitThreshold,
+		logger:                   params.Logger,
+		tenantIDHeaderName:       pCfg.TenantIDHeaderName,
+		nextConsumer:             nextConsumer,
+	}
+	return processor, nil
+}
+
+func getRateLimitServiceClient(ctx context.Context, serviceHost string, servicePort uint16,
+	timeoutMillis uint32, params component.ProcessorCreateSettings) (pb.RateLimitServiceClient, error) {
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, time.Millisecond*time.Duration(timeoutMillis))
+	defer cancel()
```

Comment:
You also do not want to cancel this context at the end of the function.

Reply:
Thanks, got this doubt. Any idea when and how to cancel the context ?

Reply:
You can add a WithShutdown option when calling NewTracesProcessor https://github.com/open-telemetry/opentelemetry-collector/blob/5f1733ad50c7c7a2c7c320ea26cf1e5cd718013c/processor/processorhelper/processor.go#L46

Reply:
Updated.